cirq-core/cirq/ops/control_values.py

```diff
@@ -0,0 +1,229 @@
+# limitations under the License.
+from typing import Collection, Optional, Sequence, Union, Tuple, List, Type, cast, TYPE_CHECKING
+
+import copy
+import itertools
+
+if TYPE_CHECKING:
+    import cirq
+
+
+def flatten(sequence):
+    def _flatten_aux(sequence):
+        if isinstance(sequence, int):
+            yield sequence
+        else:
+            for item in sequence:
+                yield from _flatten_aux(item)
+
+    return tuple(_flatten_aux(sequence))
+
+
+def _from_int(val: int) -> Tuple[Tuple[int, ...], ...]:
+    return ((val,),)
+
+
+def _from_sequence_int(vals: Sequence[int]) -> Tuple[Tuple[int, ...], ...]:
+    return tuple((val,) for val in vals)
+
+
+def _from_sequence_sequence(vals: Sequence[Sequence[int]]) -> Tuple[Tuple[int, ...], ...]:
+    return tuple(tuple(product) for product in vals)
+
+
+class ControlValues:
+    def __init__(
+        self, control_values: Sequence[Union[int, Collection[int], Type['ControlValues']]]
+    ):
+        if len(control_values) == 0:
+            self.vals = cast(Tuple[Tuple[int, ...], ...], (()))
+            self.num_variables = 0
+            self.nxt = None
+            self.itr = None
+            return
+        self.itr = None
+        self.nxt = None
+
+        if len(control_values) > 1:
+            self.nxt = ControlValues(control_values[1:])
+
+        if isinstance(control_values[0], ControlValues):
+            aux = control_values[0].copy()
+            aux.product(self.nxt)
+            self.vals, self.num_variables, self.nxt = aux.vals, aux.num_variables, aux.nxt
+            self.vals = cast(Tuple[Tuple[int, ...], ...], self.vals)
+            return
+
+        val = control_values[0]
+        if isinstance(val, int):
+            self.vals = _from_int(val)
+        elif isinstance(val, (list, tuple)):
+            if isinstance(val[0], int):
+                self.vals = _from_sequence_int(val)
+            else:
+                self.vals = _from_sequence_sequence(val)
+        else:
+            raise TypeError(f'{val} is of Unsupported type {type(val)}')
+        self.num_variables = len(self.vals[0])
+
+    def product(self, other):
+        # Cartesian product of all combinations in self x other
+        if other is None:
+            return
+        other = other.copy()
+        cur = self
+        while cur.nxt is not None:
+            cur = cur.nxt
+        cur.nxt = other
+
+    def __call__(self):
+        return self.__iter__()
+
+    def __iter__(self):
+        nxt = self.nxt if self.nxt else lambda: [()]
+        if self.num_variables:
+            self.itr = itertools.product(self.vals, nxt())
+        else:
+            self.itr = itertools.product(*(), nxt())
+        return self.itr
+
+    def copy(self):
+        if self.num_variables == 0:
+            new_copy = ControlValues([])
+        else:
+            new_copy = ControlValues(
+                [
+                    copy.deepcopy(self.vals),
```

Comment:
Can this be done without the copy module?

Reply:
If that's the case, you shouldn't need *any* copy since tuples are immutable. If you remove the copy completely does something break?